### ISSUE TITLE
An alternative to Singular on Windows OS?

### ISSUE THREAD

**Issue by PoeticExposition** [xiaxueqaq/GeometricCAD#1]
> For primary decomposition and minimal primes computation, we interface to Singular [[DGPS21](https://www.singular.uni-kl.de:8003)] because there is no native support of primary decomposition in Mathematica so far. 

> The function `minAssGTZ` in Singular fails to compute the minimal primes of ideals in FIS-3 and FIS-6 in 1 hour, but Maple can carry the same computation in less than half a minute.

Considering that Singular is _not ready-to-use_ on Windows and Maple itself is not a non-commercial product, may you contemplate invoking Magma's [`RadicalDecomposition`](http://magma.maths.usyd.edu.au/magma/handbook/text/1279#14553) instead? Regretfully, I do not possess the most recent licence for it, hence I am unable to test it [offline](https://magma.maths.usyd.edu.au/calc). 
However, there exists a ready-made function [primedec](http://www.math.kobe-u.ac.jp/OpenXM/Current/doc/asir2000/html-en/man/man_58.html#index-primedec&:~:text=noro_pd.rr) in the latest release of Risa/Asir, whose free (binary) distribution for Windows is available [here](http://air.s.kanazawa-u.ac.jp/~ohara/asir/index.html).  If ``GeometricCAD`Private`PrimeDecomposition`` is changed into 
```Mathematica
(* Suppose that Risa/Asir's installation directory is "C:\\Program Files\\asir\\". *)
PrimeDecomposition[f_List, {x__Symbol?AtomQ}] := 
 Module[{str, ans}, 
  ans /; MatchQ[{OrderlessPatternSequence[x]}][Variables[f]] \[And] 
    AllTrue[Curry[PolynomialExpressionQ][{x}]][f] \[And] 
    VectorQ[ans = 
      ImportString[
       StringReplace["[" \[Rule] "List["][
        Last[WithCleanup[str = File[CreateTemporary[]], 
          Import["!C:/Program Files/asir/bin/asir.exe -f " <> 
            Export[str, {"load(\"noro_pd.rr\")$", 
              StringTemplate[
                "noro_pd.prime_dec2([<*StringTake[ToString[\
TemplateSlot[1],InputForm],{2,-2}]*>],[<*StringTake[ToString[\
TemplateSlot[2],InputForm],{2,-2}]*>]);"][f, {x}]}, "Lines"], 
           "Lines"], DeleteFile[str]; 
          FileSystemScan[DeleteFile, File[$TemporaryDirectory], 
           FileNameForms \[Rule] "asir*"]]]], "WL"], VectorQ]] 
(* 
 * When performance is a concern, manually manage the external session is better.
 * To put it differently: `proc = StartProcess["C:/Program Files/asir/bin/asir.exe"];`.
 * This may save a substantial amount of time, whereas I have no time to modify them.
 *)
```
the “GeometricCAD” package will be capable of working **without calling WSL**. 
The only problem is that I hate to install the Windows Subsystem for Linux, so I cannot check if the results returned by Risa/Asir correspond with those returned by Singular (although Singular can be used online). Might you please verify whether the output of the two is identical/equivalent?

**Comment by xiaxueqaq** [comment]
Thanks for your interest in our work.

> Considering that Singular is not ready-to-use on Windows and Maple itself is not a non-commercial product, may you contemplate invoking Magma's [RadicalDecomposition](http://magma.maths.usyd.edu.au/magma/handbook/text/1279#14553) instead?

Neither my institution nor I have a license for Magma. 

> The only problem is that I hate to install the Windows Subsystem for Linux

I believe there is a workaround. Before WSL was released, Singular could be installed through cygwin or MSYS2. In fact, the Singular interface in Mathematica was originally designed for cygwin ([Link](http://www.kauers.de/software.html)).

> Might you please verify whether the output of the two is identical/equivalent?

In principle, the minimal primes of an ideal is unique (unlike primary decomposition). So if the algorithm is properly implemented, then the result should be equivalent.

---------------------------------------------------------------------------------------------------------------------------------------

PS: Actually, the prime decomposition in the algorithm can be avoided. However, this might complicate the paper and it is unconvenient to state some geometric results without the integral condition. It is still unclear that whether a prime decomposition is favorable in computation. On the one hand, prime ideals are nicer and this decomposition can drop the dimension or the degree of an ideal. On the other hand, decomposition itself is considered to be computationally expensive. This is certainly a topic worth exploring. However, our current focus is more on the theoretical aspect (there are two reasons for this, 1. this geometrical theory can be further developed, 2. I am doing this single-handed and heavy engineering/experiments require much more time than I have.)

**Comment by PoeticExposition** [comment]
Thanks. I am concentrating on algebraic inequalities as well as analytic inequalities at present, so I hope your novel theory can assist to prove them.  Unfortunately, the meaning of input and output is more or less undocumented and not yet clear to me. Even though the Algorithm 7 in the referenced paper mentions that the returned result is a cylindrical algebraic decomposition, I find that it appears to be quite distinct with the [built-in one in Mma](https://www.wolfram.com/language/12/algebraic-computation/compute-solution-set-topological-properties.html?view=all#:~:text=CylindricalDecomposition) (namely, `CylindricalDecomposition` and `GenericCylindricalDecomposition`). 
For instance, the translation of [qepcad-fis-1.txt](https://github.com/xiaxueqaq/GeometricCAD/blob/main/Experiments/qepcad-fis-1.txt#L27)'s 
```QEPCAD B
=======================================================
Enter an informal description  between '[' and ']':
[FIS-ex-1]
Enter a variable list:
(a,b,c,x,y)
Enter the number of free variables:
3
Enter a prenex formula:
(A x)(A y)[[ b^2 (x-c)^2 + a^2 y^2 = a^2 b^2 ] ==> [x^2+y^2<=1]].


=======================================================
```
 seems to be `ForAll[{x, y}, b^2*(x - c)^2 + a^2*y^2 \[LongEqual] a^2*b^2, x^2 + y^2 \[LessSlantEqual] 1]`, but what do 
```Mathematica
(*FIS-1-eqs.nb*)
GeometricCylindricalDecompositionEqs[{{{x^2 + y^2 - 1}, 1}, {{}, 
    1}}, {b^2  (x - c)^2 + a^2  y^2 - a^2  b^2}, {a, b, c, x, y}, True]
(*FIS-1-full.nb*)
GeometricCylindricalDecomposition[{{{b^2  (x - c)^2 + a^2  y^2 - 
      a^2  b^2}, 1}, {{x^2 + y^2 - 1}, 1}}, {a, b, c, x, y}, True]
```
 mean? I'm confused—do both of them mean `{a, b, c, x, y} /. SemialgebraicComponentInstances[b^2*(x - c)^2 + a^2*y^2 \[Equal] a^2*b^2 \[Implies] x^2 + y^2 \[LessEqual] 1, {a, b, c, x, y}]`?

**Comment by xiaxueqaq** [comment]
> Thanks. I am concentrating on algebraic inequalities as well as analytic inequalities at present, so I hope your novel theory can assist to prove them. Unfortunately, the meaning of input and output is more or less undocumented and not yet clear to me. Even though the Algorithm 7 in the referenced paper mentions that the returned result is a cylindrical algebraic decomposition, I find that it appears to be quite distinct with the [built-in one in Mma](https://www.wolfram.com/language/12/algebraic-computation/compute-solution-set-topological-properties.html?view=all#:~:text=CylindricalDecomposition) (namely, `CylindricalDecomposition` and `GenericCylindricalDecomposition`). For instance, the translation of [qepcad-fis-1.txt](https://github.com/xiaxueqaq/GeometricCAD/blob/main/Experiments/qepcad-fis-1.txt#L27)'s
> 
> ```
> =======================================================
> Enter an informal description  between '[' and ']':
> [FIS-ex-1]
> Enter a variable list:
> (a,b,c,x,y)
> Enter the number of free variables:
> 3
> Enter a prenex formula:
> (A x)(A y)[[ b^2 (x-c)^2 + a^2 y^2 = a^2 b^2 ] ==> [x^2+y^2<=1]].
> 
> 
> =======================================================
> ```
> 
> seems to be `ForAll[{x, y}, b^2*(x - c)^2 + a^2*y^2 \[LongEqual] a^2*b^2, x^2 + y^2 \[LessSlantEqual] 1]`, but what do
> 
> ```mathematica
> (*FIS-1-eqs.nb*)
> GeometricCylindricalDecompositionEqs[{{{x^2 + y^2 - 1}, 1}, {{}, 
>     1}}, {b^2  (x - c)^2 + a^2  y^2 - a^2  b^2}, {a, b, c, x, y}, True]
> (*FIS-1-full.nb*)
> GeometricCylindricalDecomposition[{{{b^2  (x - c)^2 + a^2  y^2 - 
>       a^2  b^2}, 1}, {{x^2 + y^2 - 1}, 1}}, {a, b, c, x, y}, True]
> ```
> 
> mean? I'm confused—do both of them mean `{a, b, c, x, y} /. SemialgebraicComponentInstances[b^2*(x - c)^2 + a^2*y^2 \[Equal] a^2*b^2 \[Implies] x^2 + y^2 \[LessEqual] 1, {a, b, c, x, y}]`?

The definition of cylindrical algebraic decomposition is given in Definition 4.15, it is quite standard, as it is almost the same as in Basu, Pollack and Roy. However, in real applications, usually the description of the cells (i.e. sections and bands) is not important. Also the description is often long and it takes more computation to get a concrete quantifier-free representation in polynomials (not just root functions). In practise, the sample point representation (a sample point in each cell) is more easy to use and common. Our output is actually the family of sample points (each one represents a cell). 

For example, in `FIS-1-eqs`, the result is a family of cells of $V(b^2  (x - c)^2 + a^2  y^2 - a^2  b^2)$ such that $x^2+y^2-1$ is sign-invariant in each cell. And in `FIS-1-full`, the result is a family of cells in 5-dimensional real affine space such that both polynomials are sign-invariant in each cell. Since the sign of these polynomials in a cell can be read off from a sample point of that cell, the sample point representation is enough.

**Comment by PoeticExposition** [comment]
Oh, I get it. As opposed to the sample point representation, an explicit cell description is often more practical for me when working on algebraic inequalities with undetermined parameters (e.g., a generalized contest problem `Exists[{y, z, w, x}, Element[y | z | w | x, PositiveReals] && (1 + w)^2*x^2 == (1 - w)^2 && y*z*w + y*z*x + y*w*x + z*w*x == y + z + w + x, u == 3*w/(1 + w^2) + 4*x/(1 + x^2) + 5*v*(y/(1 + y^2) + z/(1 + z^2))]` which Mma is unable to solve within a reasonable time) unless I just need to prove an algebraic inequality without any free variables. Anyway, thanks for your careful commentary. 
(Incidentally, if I understand correctly,  the `SingularMinAssGTZ` in `AlternativeProj2` should be replaced with `PrimeDecomposition`.)
